File: app/strategy_engine/base_strategy.py
from abc import ABC
from abc import abstractmethod
from typing import Dict, Type

class BaseStrategy(ABC):

    _registry: Dict[str, Type['BaseStrategy']] = {}

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        if hasattr(cls, 'strategy_name'):
            BaseStrategy._registry[cls.strategy_name] = cls

    @abstractmethod
    async def generate_signal(
        self,
        symbol: str,
        price: float
    ):
        pass

    @classmethod
    def create(cls, strategy_name: str, **kwargs) -> 'BaseStrategy':
        if strategy_name not in cls._registry:
            raise ValueError(f"Strategy '{strategy_name}' not found in registry. Available strategies: {list(cls._registry.keys())}")
        return cls._registry[strategy_name](**kwargs)

File: backend/src/trading/strategies/baseStrategy.ts
export abstract class BaseStrategy {
  static registry: { [key: string]: new (...args: any[]) => BaseStrategy } = {};

  abstract generateSignal(symbol: string, price: number): Promise<void>;

  static register(name: string, constructor: new (...args: any[]) => BaseStrategy) {
    this.registry[name] = constructor;
  }

  static create(name: string, config?: any): BaseStrategy {
    const StrategyClass = this.registry[name];
    if (!StrategyClass) {
      throw new Error(`Strategy '${name}' not found in registry. Available: ${Object.keys(this.registry)}`);
    }
    return new StrategyClass(config);
  }
}
export default BaseStrategy;
